R/zchunk_LA124.heat.R

```diff
@@ -40,70 +40,175 @@ module_energy_LA124.heat_DISABLED <- function(command, ...) {
+      mutate(sector = sub("out_", "", sector)) %>%
+      left_join(enduse_fuel_aggregation, by="fuel")  %>%
+      select(-fuel, -electricity, -bld, -industry, -trn)%>%
+      rename(fuel=heat) %>%
+      gather(key=year, value=value, -fuel, -sector, -GCAM_region_ID) %>%
+      group_by(fuel, sector, GCAM_region_ID, year) %>%
+      summarise(value=sum(value)) %>%
+      filter(fuel != is.na(fuel)) %>%
+      mutate(year = sub("X", "", year)) %>%
+      mutate(year = as.integer(year)) -> L124.out_EJ_R_heatfromelec_F_Yh
+
+
+    #Secondary output coefficients on heat produced by main activity CHP plants
+    L1231.out_EJ_R_elec_F_tech_Yh %>%
+      filter(GCAM_region_ID %in% heat_regionIDs) %>%
+      filter(technology %in% calibrated_techs$technology[calibrated_techs$secondary.output == "heat"]) %>%
+      gather(key=year, value=value, -fuel, -sector, -GCAM_region_ID, -technology) %>%
+      mutate(year = sub("X", "", year)) %>%
+      mutate(year = as.integer(year)) %>%
+      left_join(L124.out_EJ_R_heatfromelec_F_Yh %>%
+                  rename(value_elec=value) %>%
+                  rename(temp=sector), by=c("GCAM_region_ID", "fuel", "year")) %>%
+      mutate(value=value_elec/value) %>%
+      select(-value_elec, -temp)-> L124.heatoutratio_R_elec_F_tech_Yh
+
+    #Reset missing and infinite values (applicable for CC in the base years) to 0
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      mutate(value=if_else(is.na(value)|is.infinite(value), 0, value))-> L124.heatoutratio_R_elec_F_tech_Yh
+
+    #Drop all rows where all years are 0
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      group_by(GCAM_region_ID, technology, sector, fuel) %>%
+      summarise(sum=sum(value)) %>%
+      filter(sum==0) -> all_years_0
+
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      left_join(all_years_0) %>%
+      mutate(sum=if_else(is.na(sum), 2, sum)) %>% #2 was a random, nonzero choice
+      filter(sum!=0)  %>%
+      select(-sum) -> L124.heatoutratio_R_elec_F_tech_Yh
+
+    # This is complicated. Some historical regions / years have 0 district heat output (all fuels). This is fine unless there is heat output from
+    # IEA's MAINCHP flow, which we model in the electric sector with a secondary output of heat to the district heat sector. In GCAM this will
+    # cause solution failure as the district heat supply will be 0. In order to avoid this outcome we need a nominal amount of heat produced
+    # in the zero years. The code below sets an exogenous year as the one to use to extract fuel shares for each region. In all 0 region / years,
+    # output is equal to the output in the fuel-share year times 1e-3 to avoid throwing off the energy balances.
+    fuel_share_year <- "2010"
+    L124.in_EJ_R_heat_F_Yh %>%
+      group_by(sector, GCAM_region_ID, year) %>%
+      summarise(value=sum(value)) %>%
+      mutate(value=if_else(value==0, 1e-3, value)) %>%
+      mutate(value=if_else(value!=1e-3, 0, value)) -> L124.mult_R_heat_Yh
```

Comment:
These last two lines are very confusing.  As best I can tell, the idea is to zero out everything that is nonzero, and set everything that was initially zero to 1e-3.  That seems a very strange thing to do.  If that really is what we want to do, then we should probably make it explicit:  `mutate(value = if_else(value == 0.0, 1e-3, 0.0))`.